keymapper/injection/keycode_mapper.py

```diff
@@ -234,11 +235,24 @@ def __init__(self, context, source, forward_to):
                         f'but got {key}'
                     )
 
-    def macro_write(self, ev_type, code, value):
-        """Handler for macros."""
+    def macro_write(self, ev_type, code, value, device=False):
+        """Handler for macros.
+        When device is not False, pass on to separate function for clarity"""
+        if not type(device)==bool:
+            self.device_write(ev_type, code, value, device)
+            return
         self.context.uinput.write(ev_type, code, value)
         self.context.uinput.syn()
 
+    def device_write(self, ev_type, code, value, device):
+        """Handler for source devices."""
+        #TODO: when dev parameter, locate that specific device
+        for source in self.context.sources:
+            if code in source.capabilities().get(ev_type, []):
+                source.write(ev_type, code, value)
+                return
+        logger.warning('No device with capability %s', ecodes.bytype[ev_type][code])
```

Comment:
as long as `device` is only a boolean lets name it accordingly (`to_device`) and not pass it to `device_write`, since it is not used in there.

I'd also remove the `#TODO` and instead open a issue and label it with "feature request" on github